### PR TITLE
TMC: fix translation of try-with body

### LINKED ISSUE #12957: Tail-mod-cons produces erroneous results when the TMC call is inside an exception handler.
Hello,

I believe I ran into a bug while playing around with the tail-modulo-cons extension. This code compiles without warnings with OCaml 5.1.1 (`tailcall` annotations added for clarity, no warnings either if you remove them):
```ocaml
let[@tail_mod_cons] rec test_1 = function
  | [] -> []
  | hd :: tl ->
     try int_of_string hd :: (test_1[@tailcall false]) tl
     with Failure _ -> -1 :: (test_1[@tailcall true]) tl

let result = test_1 ["1"; "2"; "foo"; "3"; "4"]
```
But the resulting `result` value is structured as such:
```
Block with size 2:
  Value 1
  Block with size 2:
    Value 2
    Block with size 2:
      Value -1
      Value 24029

[1; 2; -1; <???>]
```
Notice how the list is shorter than it should be, and how the last node is a strange value in place of either a block or the value 0. Calling `List.iter` on this list even causes a segmentation fault when the code is compiled with `ocamlc`. Removing the `tail-mod-cons` annotation does make the code produce correct results.

### PR DESCRIPTION
**Pull request by nojb**
I believe this fixes the erroneous compilation in #12957.

Fixes #12957

### CODE REVIEW

**Review by gasche — COMMENTED** [review]
I agree that this fix is correct (it is slightly better than the fix I came up with in a hurry, congrats), but I think we should audit the transformation to see if I made a similar mistake in other places.

File: lambda/tmc.ml [hunk 1]
```diff
@@ -644,8 +644,8 @@ let rec choice ctx t =
            not in tail-call position (after it returns
            we need to remove the exception handler),
            so it is not transformed here *)
```
Comment (gasche) [inline lambda/tmc.ml]: Nit: the "so it is not transformed here" part of that comment was the bug, and it should now be removed.

**Comment by gasche** [comment]
Audit in progress: I find the handling of `l1` below dubious for similar reasons:

```ocaml
    | (Psequand | Psequor) as shortcutop ->
        let l1, l2 = match primargs with
          |  [l1; l2] -> l1, l2
          | _ -> invalid_arg "choice_prim" in
        let l1 = traverse ctx l1 in
        let+ l2 = choice ctx ~tail l2 in
        Lprim (shortcutop, [l1; l2], loc)
```

**Comment by gasche** [comment]
Repro case for a `Pseqand` bug:

```ocaml
let[@tail_mod_cons] rec f () = Some (g ())
and[@tail_mod_cons] g () = false && true
```

**Comment by gasche** [comment]
I see two ways to fix the `Pseqand | Pseqor` case:

1. We could translate them into the corresponding `if..then..else..` form on the fly. (This desugaring would only be applied on occurrences in *return position* in TMC functions, where it is arguably the right thing to do.)
2. Or we could *not* try to be clever here, remove the special handling and move them to the "in common cases we just return" case below.

I think that (2) is a good move here. I do remember writing recursive functions relying on the shortcut behavior of `&&` or `||`, with recursive calls in tail position. But I don't see any natural example where doing this is beneficial as part of a TMC transformation, so I would suggest not paying for the implementation complexity here. If someone comes up with a good use-case for a clever handling of shortcut operators we can always revert to (1) later.

@nojb would you include the repro case and a fix along (2) in your PR? (Should I write another repro case for `Pseqor`, or will you have fun building one yourself?)

**Comment by gasche** [comment]
One way to think about the bug is as follows. A term constructor has subterms, on which we call either `traverse` or `choice`. We must ensure that, in all execution paths in the original term, exactly one of the `choice` subterms returns a value, which is the final value of the whole subterm.

When we translate the sequence `e1; e2`, we use `traverse` on `e1` and `choice` on `e2`. Using `choice` on e1 would be incorrect, as it may then happen that two choice subterms return a value. Calling `choice` on `e1` and `traverse` on `e2` would also be incorrect, as the value of `e2`, not `e1`, will be returned.

Before this PR, the transformation of `try e1 with exn -> e2` would call `traverse` on `e1` and `choice` on `e2`. This is wrong, because the non-failing execution paths evaluate only `e1` and return its value, so it needs to be a `choice`. Doing this is okay because we only evaluate `e2` in the case where `e1` does *not* return a value (it raises instead).

**Comment by nojb** [comment]
> Repro case for a `Pseqand` bug:
> 
> ```ocaml
> let[@tail_mod_cons] rec f () = Some (g ())
> and[@tail_mod_cons] g () = false && true
> ```

This example triggers Warning 71:
```
Line 2, characters 22-40:
2 | and[@tail_mod_cons] g () = false && true
                          ^^^^^^^^^^^^^^^^^^
Warning 71 [unused-tmc-attribute]: This function is marked @tail_mod_cons
but is never applied in TMC position.
```
Is that expected?

**Comment by nojb** [comment]
> @nojb would you include the repro case and a fix along (2) in your PR? (Should I write another repro case for `Pseqor`, or will you have fun building one yourself?)

I did this (I had to disable warning 71 in the test script), thanks for the hints!

File: lambda/tmc.ml (hunk 1 above)
Comment (nojb) [reply lambda/tmc.ml]: Done, thanks.

**Review by gasche — APPROVED** [review]

**Comment by gasche** [comment]
Merged, cherry-picked in 5.2 ( a419c76651b7d6c554d63ec38440ec6faaf0eed2 ) and 4.14 ( 77f09d8f52fecbf477cbc8082bfa08013c1c3cb9 ).

**Pull request merged**